juju/juju
# [JUJU-3233] Back firewall rules with model config

Back firewall rules commands with model config

Also remove firewaller api from client since it is no longer used. Keep it in
the apiserver, to be dropped at next (minor?) release 

Add model-config items
- ssh-allow (doesn't do anything yet)
- saas-ingress-allow (previously juju-application-offer fw rule)

Use model-config internally instead of the firewall rules endpoints.

This has the side effect that the default value for
juju-application-offer & ssh firewall rule of "0.0.0.0/0" instead of returning
a NotFound error.

Remove firewaller falling back to public is juju-application-offer is
empty. Since the it can only be empty if explicitly set so

Deprecate firewall-rule commands
 
The firewall rules api has existed for a while, not being implemented.
Now that it needs to be implemented, storing this kind of model config
together in model-config makes more sense. Also, ssh and
juju-application-offer both behave quite differently, so doesn't make
sense to keep them together

## Checklist


## QA steps

Bootstrap a controller, add a model and verify `ssh-allow` & `saas-ingress-allow` have the correct default
```
$ juju bootstrap lxd lxd
$ juju add-model m
$ juju model-config ssh-allow
0.0.0.0/0,::/0
$ juju model-config saas-ingress-allow
0.0.0.0/0,::/0
$ juju list-firewall-rules
WARNING ...

Service                 Whitelist subnets
juju-application-offer  0.0.0.0/0,::/0
ssh                     0.0.0.0/0,::/0
```

Set ssh-allowlist using model-config
```
$ juju model-config ssh-allow="192.168.0.0/24"
$ juju model-config ssh-allow
192.168.0.0/24"
$ juju model-config saas-ingress-allow="192.168.1.0/24"
$ juju model-config saas-ingress-allow
192.168.1.0/24"
$ juju list-firewall-rules
WARNING ...

Service                 Whitelist subnets
juju-application-offer  192.168.1.0/24
ssh                     192.168.0.0/24
```

Using set-firewall-rule
```
$ juju set-firewall-rule ssh --allowlist "192.168.1.0/24"
WARNING This command now just sets/reads the "ssh-allowlist" model-config item and
is deprecated in favour of setting/reading that item with "juju model-config"
$ juju model-config ssh-allowlist
192.168.1.0/24"
$ juju list-firewall-rules
WARNING This command now just sets/reads the "ssh-allowlist" model-config item and
is deprecated in favour of setting/reading that item with "juju model-config"

Service                 Whitelist subnets
juju-application-offer  192.168.1.0/24
ssh                     192.168.1.0/24
```
Repeat this with juju-application-offer

### Verify a model migration

NOTE: At the moment, model migrations are failing due to a bug independent of this change, which has yet to be resolved

This can be worked around to pass QA by applying the following patch
```
index 203c2235f7..1b6a078dae 100644
--- a/apiserver/httpcontext.go
+++ b/apiserver/httpcontext.go
@@ -237,7 +237,7 @@ func (a controllerAdminAuthorizer) Authorize(authInfo httpcontext.AuthInfo) erro
                }
        }
        if a.hasPermission == nil {
-               return errors.New("no permission checker configured")
+               return nil
        }
        userTag, ok := authInfo.Entity.Tag().(names.UserTag)
        if !ok {
```

Setup a 3.1 controller/model
```
$ juju-v3.1 version
3.1.2-ubuntu-amd64
$ juju-v3.1 bootstrap lxd lxd-src
(wait)
$ juju-v3.1 add-model m
$ juju-v3.1 set-firewall-rule ssh --whitelist 192.168.0.0/24,192.168.2.0/24
$ juju-v3.1 set-firewall-rule juju-application-offer --whitelist 10.0.0.0/16
$ juju-v3.1 list-firewall-rules
Service                 Whitelist subnets
juju-application-offer  10.0.0.0/16
ssh                     192.168.0.0/24,192.168.2.0/24
```

Bootstrap a ctrl from this PR and model migrate
```
$ juju-v3.2 version
3.2-beta1-ubuntu-amd64
$ juju-v3.2 bootstrap lxd lxd-dst
(wait)
$ juju-v3.2 switch lxd-src
$ juju-v3.2 migrate m lxd-dst
Migration started with ID "86a70b5a-ac41-49c5-8a7f-ec37ed095b0b:0"
(wait)
$ juju-v3.2 models
Controller: lxd-src

Model       Cloud/Region         Type  Status     Machines  Units  Access  Last connection
controller  localhost/localhost  lxd   available         1      1  admin   just now

$ juju-v3.2 switch lxd-dst:admin/m
$ juju-v3.2 model-config ssh-allow
192.168.0.0/24,192.168.2.0/24
$ juju-v3.2 model-config saas-ingress-allow
10.0.0.0/16
```

Repeat this for a 3.0 and 2.9 source controller

### Verify cmr's still function
Verify the `cmr` integration test passes to verify firewall changes
```
./main.sh -v -p aws -c aws cmr
```

And verify that cmrs operate as specified here:
https://juju.is/docs/olm/manage-cross-model-integrations
(@wallyworld I need some help verifying this)
